Allow MCSystemInterface::ListFolderEntries() of arbitrary paths.

peter-b opened the pull request:
Sometimes it's useful to iterate over the contents of directories
other than the current working directory.

N.b. I haven't been able to build & test all of these changes, so merging this might be risky. :neutral_face:

peter-b committed:
Allow MCSystemInterface::ListFolderEntries() of arbitrary paths.

Sometimes it's useful to iterate over the contents of directories
other than the current working directory.

runrevmark commented on engine/src/mbliphone.mm:
```diff
@@ -815,10 +815,16 @@ virtual int64_t GetFileSize(void)
+
 	DIR *t_dir;
-	t_dir = opendir(".");
+	t_dir = opendir(MCStringGetCString(t_path));
```
This needs to MCAutoStringRefAsUTF8String'd

runrevmark commented on engine/src/mblandroidfs.cpp:
```diff
@@ -500,14 +496,21 @@ bool MCAndroidSystem::ResolvePath(MCStringRef p_path, MCStringRef& r_resolved)
 
 	DIR *t_dir;
-	t_dir = opendir(".");
+	t_dir = opendir(MCStringGetCString (t_path));
```
I think this also needs AsUTF8String'd

peter-b committed:
MCIPhoneSystem::ListFolderEntries(): Ensure path is UTF-8 encoded.

peter-b committed:
MCAndroidSystem::ListFolderEntries(): Ensure path is UTF-8 encoded.

runrevmark commented on engine/src/dsklnx.cpp:
```diff
@@ -1912,9 +1912,15 @@ class MCLinuxDesktop: public MCSystemInterface
         closedir(dirptr);
         *dptr = tptr;
 #endif /* MCS_getentries_dsk_lnx */
+		MCStringRef t_path;
```
This needs to be MCAutoStringRefAsSysString'd :)

peter-b committed:
MCLinuxDesktop::ListFolderEntries(): Ensure path has system encoding.

peter-b committed:
MCPosixSystem::ListFolderEntries: Ensure path has system encoding.

peter-b commented:
This patch series as it stands doesn't work because most of the platforms' implementations subsequently call:

```
struct dirent *t_dir_entry;
struct stat t_stat;
t_dir_entry = readdir(t_dir);
stat(t_dir_entry->d_name, &t_stat);
```

Obviously, the call to `stat` won't work correctly if the directory of `t_dir` isn't the current working directory.

peter-b committed:
[[ListFolderEntries]] Properly handle directories other than the CWD.

On some platforms, directory entry names are passed directly to
e.g. the stat(2) system call in order to get information such as the
file type, size and permissions.

When arbitrary directories can be listed, it is necessary to construct
a proper path, rather than assuming that the directory entry name will
be correctly resolved relative to the current working directory.

livecodefraser commented on engine/src/mblandroidfs.cpp:
```diff
@@ -115,16 +115,12 @@ bool apk_set_current_folder(MCStringRef p_apk_path)
 	MCAutoStringRef t_list;
-    MCAutoStringRef t_current_folder;
-    
-    if (!apk_get_current_folder(&t_current_folder))
```
This call is still needed when the APK folder isn't passed in

livecodefraser commented:
@livecode-vulcan reviewed okay 1c5fb2c

livecode-vulcan commented:
💙 review by @livecodefraser ok 1c5fb2c04fa469e68804d5f0d551aece3792fd42

livecode-vulcan commented:
😞 test failure 1c5fb2c04fa469e68804d5f0d551aece3792fd42
- try-linux-x86: [failure](https://vulcan.livecode.com/builders/try-linux-x86/builds/203)
- try-win-x86: [failure](https://vulcan.livecode.com/builders/try-win-x86/builds/267)
- try-ios: unknown\* try-android-armv6: [failure](https://vulcan.livecode.com/builders/try-android-armv6/builds/214)
- try-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/149)
- try-mac: unknown

peter-b commented:
@livecode-vulcan try 1c5fb2c

livecode-vulcan commented:
💥 merge conflict 1c5fb2c04fa469e68804d5f0d551aece3792fd42 → 3e4576d5d0699a30680ae4bbb230ac6a5833dfbf

peter-b committed:
Merge branch 'develop' into misc-listfolderentries

peter-b commented:
@livecode-vulcan reviewed by @runrevmark ok 596ad32

livecode-vulcan commented:
💙 review by @runrevmark ok 596ad3239b0ecce1a89e51d41e779978f8b9fada

peter-b committed:
[[ListFolderEntries]] Fix type in iPhoneOS implementation.

runrevmark commented:
@livecode-vulcan review ok f3780bd

livecode-vulcan commented:
💙 review by @runrevmark ok f3780bd61a26c71ade5282b16267d410efc4f2aa

livecode-vulcan commented:
😞 test failure f3780bd61a26c71ade5282b16267d410efc4f2aa
- try-linux-x86: [success](https://vulcan.livecode.com/builders/try-linux-x86/builds/452)
- try-win-x86: [pending](https://vulcan.livecode.com/builders/try-win-x86/builds/512)
- try-ios: [success](https://vulcan.livecode.com/builders/try-ios/builds/398)
- try-android-armv6: [failure](https://vulcan.livecode.com/builders/try-android-armv6/builds/463)
- try-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/398)
- try-mac: [success](https://vulcan.livecode.com/builders/try-mac/builds/389)

peter-b committed:
[[ListFolderEntries]] Fix incorrect variable in Android implementation.

livecodeali commented:
@livecode-vulcan review ok c5bd220

livecode-vulcan commented:
💙 review by @livecodeali ok c5bd220aa40f7b515717b97b21f3ef1468009fb4

livecode-vulcan commented:
😎 test success c5bd220aa40f7b515717b97b21f3ef1468009fb4
- try-linux-x86: [success](https://vulcan.livecode.com/builders/try-linux-x86/builds/641)
- try-win-x86: [success](https://vulcan.livecode.com/builders/try-win-x86/builds/701)
- try-ios: [success](https://vulcan.livecode.com/builders/try-ios/builds/590)
- try-android-armv6: [success](https://vulcan.livecode.com/builders/try-android-armv6/builds/654)
- try-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/586)
- try-mac: [success](https://vulcan.livecode.com/builders/try-mac/builds/579)
